File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        def permutation():
            if len(route) == len(nums):
                result.append(list(route))
                return

            for i in range(len(nums)):
                if not visited[i]:
                    visited[i] = True
                    route.append(nums[i])
                    permutation()
                    visited[i] = False
                    route.pop()


        visited = [False for i in range(len(nums))]
        result = list()
        route = list()
        permutation()

        return result

File: 0046-permutations/0046-permutations.java
class Solution {
    List<List<Integer>> result = new ArrayList<>();

    private void backtrack(List<Integer> numbers, int[] nums, Set<Integer> used) {
        if (numbers.size() == nums.length) {
            result.add(new ArrayList<>(numbers));
            return;
        }
        for (int num: nums) {
            if (used.contains(num)) {
                continue;
            }
            numbers.add(num);
            used.add(num);
            backtrack(numbers, nums, used);
            used.remove(num);
            numbers.remove(numbers.size() - 1);
        }
    }

    public List<List<Integer>> permute(int[] nums) {
        backtrack(new ArrayList<>(), nums, new HashSet<Integer>());
        return result;
    }
}